src/doc/rustc/src/profile-guided-optimization.md

```diff
@@ -0,0 +1,134 @@
+version usually works too.
+
+The `llvm-profdata` tool merges multiple `.profraw` files into a single
+`.profdata` file that can then be fed back into the compiler via
+`-Cprofile-use`:
+
+```bash
+# STEP 1: Compile the binary with instrumentation
+rustc -Cprofile-generate=/tmp/pgo-data -O -Ccodegen-units=1 ./main.rs
+
+# STEP 2: Run the binary a few times, maybe with common sets of args.
+#         Each run will create or update `.profraw` files in /tmp/pgo-data
+./main mydata1.csv
+./main mydata2.csv
+./main mydata3.csv
+
+# STEP 3: Merge and post-process all the `.profraw` files in /tmp/pgo-data
+llvm-profdata merge -o ./merged.profdata /tmp/pgo-data
+
+# STEP 4: Use the merged `.profdata` file during optimization. All `rustc`
+#         flags have to be the same.
+rustc -Cprofile-use=./merged.profdata -O -Ccodegen-units=1 ./main.rs
+```
+
+Note that setting `-Ccodegen-units=1` will produce better runtime performance
+at the expense of longer compile times; a reasonable trade-off in scenarios
+where PGO is used in the first place.
+
+### A Complete Cargo Workflow
+
+Using this feature with Cargo works very similar to using it with `rustc`
+directly. Again, we generate an instrumented binary, run it to produce data,
+merge the data, and feed it back into the compiler. Some things of note:
+
+- We use the `RUSTFLAGS` environment variable in order to pass the PGO compiler
+  flags to the compilation of all crates in the program.
+
+- We pass the `--target` flag to Cargo, which prevents the `RUSTFLAGS`
+  arguments to be passed to Cargo build scripts. We don't want the build
+  scripts to generate a bunch of `.profraw` files.
+
+- We pass `--release` to Cargo because that's where PGO makes the most sense.
+  In theory, PGO can also be done on debug builds but there is little reason
+  to do so.
+
+- It is recommended to use *absolute paths* for the argument of
+  `-Cprofile-generate` and `-Cprofile-use`. Cargo can invoke `rustc` with
+  varying working directories, meaning that `.profraw` files might show up
+  in unexpected places or that `rustc` cannot find the supplied `.profdata`
+  file. With absolute paths this is not an issue.
+
+- It is good practice to make sure that there is no left-over profiling data
+  from previous compilation sessions. Just deleting the directory is a simple
+  way of doing so (see `STEP 0` below).
+
+This is what the entire workflow looks like:
+
+```bash
+# STEP 0: Make sure there is no left-over profiling data from previous runs
+rm -rf /tmp/pgo-data
+
+# STEP 1: Build the instrumented binaries
+RUSTFLAGS="-Cprofile-generate=/tmp/pgo-data -Ccodegen-units=1" \
+    cargo build --release --target=x86_64-unknown-linux-gnu
+
+# STEP 2: Run the instrumented binaries with some typical data
+./target/x86_64-unknown-linux-gnu/release/myprogram mydata1.csv
+./target/x86_64-unknown-linux-gnu/release/myprogram mydata2.csv
+./target/x86_64-unknown-linux-gnu/release/myprogram mydata3.csv
```

Comment:
This might even be nifty as:

```sh
export RUSTFLAGS=...
cargo run --release --target x86_64-unknown-linux-gnu mydata1.csv
cargo run --release --target x86_64-unknown-linux-gnu mydata2.csv
cargo run --release --target x86_64-unknown-linux-gnu mydata3.csv

# ...
```

(basically using `cargo run` instead of reaching into the target directory)

Reply:
`cargo run` will rebuild of it is invoked with different `RUSTFLAGS` than the preceding `cargo build`, right? 

Your example is nice but I'm a bit worried about
- globally modifying `RUSTFLAGS` and
- too much stuff is happening here implicitly.

I'll think about it.

Reply:
You're right yeah, my suggestion is probably too clever by half, I wouldn't bother with it.